src/gpu/intel/jit/codegen/kernel.hpp

```diff
@@ -1154,133 +1132,155 @@ class ir_kernel_base_t : public ngen_generator_t {
             const compute::range_t &local_range, bool require_dpas,
             const debug_config_t &debug_config)
-        : base(exec_cfg, {}, debug_config)
-        , kernel_name_(kernel_name)
-        , require_dpas_(require_dpas)
-        , local_range_(local_range) {}
-
-    const ngen::NEOInterfaceHandler &neo_interface() const {
-        return elf_generator_t::interface_;
+        : exec_cfg_(exec_cfg)
+        , local_range_(local_range)
+        , interface_(exec_cfg_.hw())
+        , debug_config_(debug_config) {
+        interface_.externalName(kernel_name);
+        if (require_dpas) interface_.requireDPAS();
     }
 
+    const exec_config_t &exec_cfg() const { return exec_cfg_; }
+    const kernel_iface_t &kernel_iface() const { return kernel_iface_; }
+
     void set_kernel_iface(const kernel_iface_t &kernel_iface) {
-        base::kernel_iface_ = kernel_iface;
+        kernel_iface_ = kernel_iface;
     }
 
     void setup_interface(const stmt_t &kernel_body = stmt_t()) {
-        elf_generator_t::externalName(kernel_name_);
-        elf_generator_t::requireLocalID(3);
-        elf_generator_t::requireLocalSize();
-        elf_generator_t::requireGRF(base::exec_cfg().regs());
-        elf_generator_t::requireSIMD(base::exec_cfg().simd());
-        elf_generator_t::requireBarrier();
+        interface_.requireLocalID(3);
+        interface_.requireLocalSize();
+        interface_.requireGRF(exec_cfg_.regs());
+        interface_.requireSIMD(exec_cfg_.simd());
+        interface_.requireBarrier();
         auto setup_flags = get_setup_flags(kernel_body);
 
-        // On XeHPG hardware, there is some setup overhead when switching
-        // between dpas and non-dpas kernels. The require_dpas_ parameter is to
-        // enable better kernel chaining on this platform.
-        if (require_dpas_ || setup_flags.has_dpas)
-            elf_generator_t::requireDPAS();
-        if (setup_flags.has_send_atomics)
-            elf_generator_t::requireGlobalAtomics();
-
-        for (int i = 0; i < base::kernel_iface().nargs(); i++) {
-            auto &name = base::kernel_iface().arg_name(i);
-            auto &type = base::kernel_iface().arg_type(i);
+        if (setup_flags.has_dpas) interface_.requireDPAS();
+        if (setup_flags.has_send_atomics) interface_.requireGlobalAtomics();
+
+        for (int i = 0; i < kernel_iface_.nargs(); i++) {
+            auto &name = kernel_iface_.arg_name(i);
+            auto &type = kernel_iface_.arg_type(i);
             if (type.is_ptr()) {
-                elf_generator_t::newArgument(name,
+                interface_.newArgument(name,
                         ngen::ExternalArgumentType::GlobalPtr,
                         ngen::GlobalAccessType::Stateless);
             } else {
-                elf_generator_t::newArgument(name, to_ngen(type));
+                interface_.newArgument(name, to_ngen(type));
             }
```

Comment:
> @rjoursler Please take a look. I noticed you added compile_ctx_t in your branch

My intention here is to simplify the interface to `codegen` to break dependencies on oneDNN types. Effectively, my goal is to reduce the interface to:
```
template <typename > kernel_object
kernel_object build(build_info_t info, stmt_t s); // kernel_object can be cl_kernel, sycl::kernel, or elf_binary_t
```
With this change, `ir_kernel_t` would be moved out of `codegen/kernel.hpp` and modified to use this interface.